public/script.js

```diff
@@ -8677,366 +8706,391 @@ export async function createOrEditCharacter(e) {
+     */
+    async function animateSwipe(run_generate = false) {
+
+        //Swipe out.
+        await animateSwipeTransition(mesId, swipeRange, swipeDuration);
+
+        if (run_generate) {
+            //shows "..." while generating
+            thisMesDiv.find('.mes_text').html('...');
+            // resets the timer
+            thisMesDiv.find('.mes_timer').html('');
+            thisMesDiv.find('.tokenCounterDisplay').text('');
+            updateReasoningUI(thisMesDiv, { reset: true });
+        } else {
+            //console.log('showing previously generated swipe candidate, or "..."');
+            //console.log('onclick right swipe calling addOneMessage');
+
+            //Only scroll when swiping the last message.
+            const scroll = (mesId == chat.length - 1);
+            addOneMessage(chat[mesId], { type: 'swipe', forceId: mesId, scroll: scroll });
+
+            if (power_user.message_token_count_enabled) {
+                if (!chat[mesId].extra) {
+                    chat[mesId].extra = {};
+                }
+
+                const tokenCountText = (chat[mesId]?.extra?.reasoning || '') + chat[mesId].mes;
+                const tokenCount = await getTokenCountAsync(tokenCountText, 0);
+                chat[mesId]['extra']['token_count'] = tokenCount;
+                thisMesDiv.find('.tokenCounterDisplay').text(`${tokenCount}t`);
+            }
+        }
+
+        //Animate expanding to the new message height.
+        thisMesDiv.css('height', thisMesDivHeight);
+        expandNewMessage(thisMesDiv);
+
+
+        //Jump to the opposite side.
+        await animateSwipeTransition(mesId, -swipeRange, 0);
+
+        appendMediaToMessage(chat[mesId], thisMesDiv);
+
+        await eventSource.emit(event_types.MESSAGE_SWIPED, (mesId));
+
+        if (run_generate && !is_send_press) {
+            is_send_press = true;
+            generation = Generate('swipe');
+        } else if (parseInt(chat[mesId]['swipe_id']) !== chat[mesId]['swipes'].length) {
+            saveChatDebounced();
+        }
+
+        //Swipe in.
+        await animateSwipeTransition(mesId, 0, swipeDuration);
     }
-}
 
-/**
- * Handles the swipe to the right event.
- * @param {JQuery.Event} [_event] Event.
- * @param {object} params Additional parameters.
- * @param {string} [params.source] The source of the swipe event.
- * @param {boolean} [params.repeated] Is the swipe event repeated.
- */
-//MARK: swipe_right
-export function swipe_right(_event = null, { source, repeated } = {}) {
-    if (chat.length - 1 === Number(this_edit_mes_id)) {
+    if (mesId === Number(this_edit_mes_id)) {
         closeMessageEditor();
     }
+    if (isStreamingEnabled() && streamingProcessor) {
+        streamingProcessor.onStopStreaming();
+    }
 
     if (isHordeGenerationNotAllowed()) {
         return unblockGeneration();
     }
 
-    // Make sure ad-hoc changes to extras are saved before swiping away
-    syncMesToSwipe();
+    hideSwipeButtons();
```

Comment:
Wishlist: hide the counters here, since streamed gen will hide them anyway, so they'll be visible for a split second until the stream starts, which is a bit jarring.

```suggestion
    hideSwipeButtons({ hideCounters: true });
```

Reply:
I would prefer leave this.
Please test this on an API that takes a few seconds for the first token.
When you swipe a message, the counter will swap from 1/1 to 2/1 as the message waits for generation to start.
Then it is hidden when the generation starts streaming.
Or if the generation is canceled, it switches back to 1/1.

Reply:
I also noticed a difference with staging where swipe buttons disappear for a split second when switching between swipes that don't require generation. So _ideally_ swipe function shouldn't call hide buttons, it should be done by the generation function as it is now.

Reply:
Removing this line fixes all the problems I have with swipe buttons flickering, except the misaligned swipes counter.
So if 2/1 is a problem, hideSwipes in generate should also hide labels.

Reply:
This patch solved it all for me. What do you think?
```diff
diff --git a/public/script.js b/public/script.js
index a2fb03c1f..74bb9dc48 100644
--- a/public/script.js
+++ b/public/script.js
@@ -3577,7 +3577,7 @@ export async function Generate(type, { automatic_trigger, force_name2, quiet_pro
         }
 
         // Hide swipes if not in a dry run.
-        hideSwipeButtons();
+        hideSwipeButtons({ hideCounters: true });
         // If generated any message, set the flag to indicate it can't be recreated again.
         chat_metadata['tainted'] = true;
     }
@@ -8079,12 +8079,17 @@ export function callPopup(text, type, inputValue = '', { okButton, rows, wide, w
 /**
  * Update the swipe counter for mesId.
  * @param {Number} mesId
+ * @param {object} [options] Options
+ * @param {boolean} [options.show=true] Whether to show the counter after updating.
  */
-export async function updateSwipeCounter(mesId) {
+export async function updateSwipeCounter(mesId, { show = true } = {}) {
     const swipeCounterText = formatSwipeCounter((chat[mesId]?.['swipe_id'] + 1), chat[mesId]?.['swipes']?.length);
     const currentMessage = chatElement.children().filter(`[mesid="${mesId}"]`);
     const swipeCounter = currentMessage.find('.swipes-counter');
-    swipeCounter.text(swipeCounterText).show();
+    swipeCounter.text(swipeCounterText);
+    if (show) {
+        swipeCounter.show();
+    }
 }
 
 /**
@@ -8909,7 +8913,8 @@ export async function swipe(_event, direction, { source, repeated, message = cha
         await animateSwipeTransition(mesId, swipeRange, swipeDuration);
 
         if (run_generate) {
-            await updateSwipeCounter(mesId);
+            hideSwipeButtons({ hideCounters: true });
+            await updateSwipeCounter(mesId, { show: false });
             //shows "..." while generating
             thisMesDiv.find('.mes_text').html('...');
             // resets the timer
@@ -8970,8 +8975,6 @@ export async function swipe(_event, direction, { source, repeated, message = cha
         return unblockGeneration();
     }
 
-    hideSwipeButtons();
-
     //If the swipe is not being deleted.
     if (source != 'delete') {
```

Reply:
>  So if 2/1 is a problem, hideSwipes in generate should also hide labels.

I prefer that 2/1 is shown to the user.

Is there any a scenario where swipe counters need to be updated in advance without being shown?
I do not understand when this could be needed.
```javascript
export async function updateSwipeCounter(mesId, { show = true } = {}) {
```

> it should be done by the generation function as it is now.

How's this?

https://github.com/SillyTavern/SillyTavern/pull/4610/commits/d9b046a0ea7697cdd4cb69296dfb4176845fafe5

Reply:
> How's this?
> 
> [d9b046a](https://github.com/SillyTavern/SillyTavern/commit/d9b046a0ea7697cdd4cb69296dfb4176845fafe5)

I originally used `hideSwipeButtons` in `swipe` to set `isSwipingAllowed` to false earlier.
Otherwise, holding the arrow key caused problems.
Setting ` isSwipingAllowed = false;` directly works.

Reply:
Click registration on buttons was a bit broken, I fixed it.

Reply:
> Click registration on buttons was a bit broken, I fixed it.

Thank you.